### ISSUE TITLE
spire-agent memory leak

### ISSUE THREAD

**Issue by scotte** [spiffe/spire#843]
To do some load testing, I am using a go client to the spire-agent workload API that sits in a loop, connecting to the workload API to fetch an SVID, then it closes the stream and sleeps for a configurable period of time.

When running this client, I observed spire-agent memory utilization to grow continuously until the system ran out of memory. With a 100ms sleep time, the spire-agent would grow to consume about 3.5GB within a minute or so. Killing off the client does not release any memory, nor does deleting workload registration entries for the client workload.

Here's a heap dump of spire-agent after running for a few seconds and killing the client:

```
      flat  flat%   sum%        cum   cum%
  208.56MB 75.38% 75.38%   208.56MB 75.38%  bytes.(*Buffer).String
   59.13MB 21.37% 96.75%   272.18MB 98.37%  github.com/spiffe/spire/pkg/agent/manager/cache.(*subscribers).add
       4MB  1.45% 98.19%        4MB  1.45%  bytes.makeSlice
       2MB  0.72% 98.92%        2MB  0.72%  github.com/spiffe/spire/pkg/common/selector.(*set).Add
    1.50MB  0.54% 99.46%     1.50MB  0.54%  strings.explode
    0.50MB  0.18% 99.64%     4.50MB  1.63%  bytes.(*Buffer).grow
    0.50MB  0.18% 99.82%        4MB  1.45%  github.com/spiffe/spire/pkg/common/selector.powerSet
         0     0% 99.82%     4.50MB  1.63%  bytes.(*Buffer).WriteString
         0     0% 99.82%   272.18MB 98.37%  github.com/spiffe/spire/pkg/agent/endpoints/workload.(*Handler).FetchX509SVID
         0     0% 99.82%   272.18MB 98.37%  github.com/spiffe/spire/pkg/agent/manager.(*manager).SubscribeToCacheChanges
```

**Comment by scotte** [comment]
![image](https://user-images.githubusercontent.com/188352/55842013-95941a80-5ae6-11e9-9a6b-e472642a527f.png)

**Comment by scotte** [comment]
While a client that makes a lot of new requests to the workload API clearly exacerbates this issue, memory usage slowly grows even when a small number (5-12) of clients streaming on the same workload API connection. Additionally, even with no active clients, memory usage of spire-agent continues to grow until (in my case, a k8s cluster) the system runs low on memory and liveness probes fail, recycling the container. When spire-agent restarts (with no active workload API clients), the issue persists, with memory usage slowly increasing.

Here's a typical example (after a spire-agent restart with no active workload API clients and 2 registration entries) - this is `ps` output every 60 seconds:
```
  PID    VSZ    RSS   %MEM CMD
 7848 2481324 2403532 60.2 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2481644 2411428 60.4 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2481644 2412220 60.4 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2481964 2429600 60.9 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2481964 2430648 60.9 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2550128 2447404 61.3 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2550128 2448096 61.3 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2550192 2466272 61.8 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2550192 2466796 61.8 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2550768 2484440 62.2 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2550768 2485492 62.3 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2551344 2502856 62.7 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2551344 2503908 62.7 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2619316 2521324 63.2 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
 7848 2619316 2522512 63.2 /opt/spire/bin/spire-agent run -config /run/spire/config/spire-agent.conf
```

**Comment by azdagron** [comment]
Based on the heap profile you collected, i believe this has to do with memory retention in the agent cache. I've implemented a fix. Any chance you could give it a whirl?

**Comment by scotte** [comment]
Yep, I'll try it right now. Thanks @azdagron !

**Comment by evan2645** [comment]
Fixed in #845
